### PR TITLE
store: use treap to manage pending slots

### PR DESCRIPTION
**Pull request by yhzhangjump**

### CODE REVIEW

**Review by lidatong — COMMENTED** [review]
good start

fix lifecycle methods (join, leave, delete, etc.), layout macro for footprint, define treap / pool in `.h` to avoid the `void *` thing` also i would keep the lock separate and inline the treap fields so

struct fd_pending_slot {
  ulong slot;
  ulong weight;

  /* The treap fields */
  ushort parent;
  ushort left;
  ushort right;
  ushort prio;
  ushort next;
  ushort prev;
};

File: src/app/fdctl/run/tiles/fd_store_int.c [hunk 1]
```diff
@@ -433,21 +428,32 @@ after_credit( void *             _ctx,
...
+    int store_slot_prepare_mode = fd_store_slot_prepare( ctx->store, ele->slot, &repair_slot, &block, &block_sz );
 
-    ulong slot = repair_slot == 0 ? i : repair_slot;
-    FD_LOG_DEBUG(( "store slot - mode: %d, slot: %lu, repair_slot: %lu", store_slot_prepare_mode, i, repair_slot ));
+    ulong slot = repair_slot == 0 ? ele->slot : repair_slot;
+    FD_LOG_DEBUG(( "store slot - mode: %d, slot: %lu, repair_slot: %lu", store_slot_prepare_mode, ele->slot, repair_slot ));
     fd_store_tile_slot_prepare( ctx, store_slot_prepare_mode, slot );
+
+    iter = fd_pending_slots_treap_fwd_iter_next( iter, pending_slots->pool );
+    fd_pending_slots_treap_ele_remove( pending_slots->treap, ele, pending_slots->pool );
```
Comment (lidatong) [inline src/app/fdctl/run/tiles/fd_store_int.c:455]: is it safe to remove here? please verify

File: src/disco/store/test_pending_slots.c [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    FD_TEST( ele->time == now + (long)ele->slot - (long)lo_wmark );
+    iter = fd_pending_slots_treap_fwd_iter_next( iter, pending_slots->pool );
+  }
+  FD_TEST( iterated == FD_BLOCK_MAX / 2 );
+
+  FD_LOG_NOTICE(("Insert %lu new elements to pending_slots", FD_BLOCK_MAX / 2));
+  for (ulong i = 0; i < FD_BLOCK_MAX / 2; i++) {
+    fd_pending_slots_add( pending_slots, FD_BLOCK_MAX + i, 0 );
+  }
+
```
Comment (lidatong) [inline src/disco/store/test_pending_slots.c:91]: add test that includes removing as you iterate as well as you do in the replay tile

**Review by lidatong — CHANGES_REQUESTED** [review]

File: src/disco/store/fd_pending_slots.c [hunk 3]
```diff
@@ -13,27 +14,25 @@ fd_pending_slots_new( void * mem, ulong lo_wmark ) {
...
 
-  laddr += sizeof(long) * FD_PENDING_MAX;
+  laddr                = fd_ulong_align_up( laddr, fd_pending_slots_pool_align() );
+  pending_slots->pool  = fd_pending_slots_pool_join( fd_pending_slots_pool_new( (void*)(laddr), FD_BLOCK_MAX ) );
+  laddr               += fd_pending_slots_pool_footprint( FD_BLOCK_MAX );
 
-  FD_TEST( laddr == (ulong)mem + footprint );
+  FD_TEST( laddr == (ulong)mem + fd_pending_slots_footprint() );
 
+  fd_pending_slots_treap_seed( pending_slots->pool, FD_BLOCK_MAX, seed );
```
Comment (lheeger-jump) [inline src/disco/store/fd_pending_slots.c:32]: It is unclear if we should generate a random number on re-insert. For correctness it is fine but for (computational) complexity maybe not.

File: src/disco/store/fd_pending_slots.c [hunk 4]
```diff
@@ -1,8 +1,9 @@
 #include "fd_pending_slots.h"
+#include <limits.h>
 
 void *
-fd_pending_slots_new( void * mem, ulong lo_wmark ) {
-
+fd_pending_slots_new( void * mem,
+                      ulong seed ) {
```
Comment (lidatong) [inline src/disco/store/fd_pending_slots.c:6]: no newline

File: src/disco/store/fd_pending_slots.c [hunk 5]
```diff
@@ -13,27 +14,24 @@ fd_pending_slots_new( void * mem, ulong lo_wmark ) {
...
+  ulong                laddr         = (ulong)(mem) + sizeof( fd_pending_slots_t );
 
-  laddr += sizeof(long) * FD_PENDING_MAX;
+  laddr                = fd_ulong_align_up( laddr, fd_pending_slots_treap_align() );
+  pending_slots->treap = fd_pending_slots_treap_join( fd_pending_slots_treap_new( (void *)(laddr), FD_BLOCK_MAX ) );
+  laddr               += fd_pending_slots_treap_footprint( FD_BLOCK_MAX );
 
-  FD_TEST( laddr == (ulong)mem + footprint );
+  laddr                = fd_ulong_align_up( laddr, fd_pending_slots_pool_align() );
+  pending_slots->pool  = fd_pending_slots_pool_join( fd_pending_slots_pool_new( (void*)(laddr), FD_BLOCK_MAX ) );
```
Comment (lidatong) [inline src/disco/store/fd_pending_slots.c:28]: void * missing space

File: src/disco/store/fd_pending_slots.c [hunk 6]
```diff
@@ -73,127 +85,50 @@ fd_pending_slots_delete( void * pending_slots ) {
     return NULL;
   }
 
+  fd_pending_slots_t * _pending_slots = pending_slots;
+  fd_pending_slots_treap_delete( fd_pending_slots_treap_leave( _pending_slots->treap ) );
+  fd_pending_slots_pool_delete( fd_pending_slots_pool_leave( _pending_slots->pool ) );
```
Comment (lidatong) [inline src/disco/store/fd_pending_slots.c:90]: why is leave called again in delete?

File: src/disco/store/fd_pending_slots.c [hunk 7]
```diff
@@ -73,127 +85,50 @@ fd_pending_slots_delete( void * pending_slots ) {
...
-    /* Grow up */
-    if( (long)(slot - pending_slots->start) > (long)FD_PENDING_MAX )
-      FD_LOG_ERR(( "pending queue overrun: start=%lu, end=%lu, new slot=%lu", pending_slots->start, pending_slots->end, slot ));
-    pending[slot & FD_PENDING_MASK] = when;
-    for( ulong i = pending_slots->end; i < slot; i++ ) {
-      /* Zero fill */
-      pending[i & FD_PENDING_MASK] = 0;
+                      long when) {
+  fd_pending_slots_treap_ele_t *ele = fd_pending_slots_treap_ele_query( pending_slots->treap, slot, pending_slots->pool );
+  if ( ele ) {
```
Comment (lidatong) [inline src/disco/store/fd_pending_slots.c:99]: no space after if and use FD_LIKELY / UNLIKELY as appropriate

File: src/disco/store/fd_pending_slots.c [hunk 8]
```diff
@@ -73,127 +85,50 @@ fd_pending_slots_delete( void * pending_slots ) {
...
+    if ( ele->time > when ) ele->time = when;
+  } else {
+    if ( fd_pending_slots_pool_free( pending_slots->pool )==0 ) {
```
Comment (lidatong) [inline src/disco/store/fd_pending_slots.c:102]: see above

File: src/disco/store/fd_pending_slots.c [hunk 9]
```diff
@@ -73,127 +85,50 @@ fd_pending_slots_delete( void * pending_slots ) {
...
 
-  } else {
-    /* Update in place */
-    long * p = &pending[slot & FD_PENDING_MASK];
-    if( 0 == *p || *p > when )
-      *p = when;
+    ele       = fd_pending_slots_pool_ele_acquire( pending_slots->pool );
+    ele->slot = slot;
+    ele->time = when;
+    ele->prio = fd_ulong_hash( ( pending_slots->total_insert++ ) ^ pending_slots->seed );
```
Comment (lidatong) [inline src/disco/store/fd_pending_slots.c:109]: fd_rng

File: src/disco/store/fd_pending_slots.c [hunk 10]
```diff
@@ -73,127 +85,50 @@ fd_pending_slots_delete( void * pending_slots ) {
...
-  if( pending_slots->start == pending_slots->end ) {
-    return LONG_MAX;
-  } else if( slot < pending_slots->start ) {
-    return LONG_MAX;
-  } else if( slot >= pending_slots->end ) {
-    return LONG_MAX;
-  } else {
-    return pending_slots->pending[ slot & FD_PENDING_MASK ];
-  }
+  fd_pending_slots_treap_ele_t *ele = fd_pending_slots_treap_ele_query( pending_slots->treap, slot, pending_slots->pool );
```
Comment (lidatong) [inline src/disco/store/fd_pending_slots.c:117]: space after *

File: src/disco/store/fd_pending_slots.c [hunk 11]
```diff
@@ -73,127 +85,50 @@ fd_pending_slots_delete( void * pending_slots ) {
...
+  return ele ? ele->time : LONG_MAX;
 }
 
 void
 fd_pending_slots_set_lo_wmark( fd_pending_slots_t * pending_slots,
-                               ulong slot ) {
-  pending_slots->lo_wmark = slot;
+                               ulong lo_wmark ) {
+  fd_pending_slots_treap_fwd_iter_t iter;
+  for ( iter = fd_pending_slots_treap_fwd_iter_init( pending_slots->treap, pending_slots->pool );
```
Comment (lidatong) [inline src/disco/store/fd_pending_slots.c:125]: declare inside loop?

File: src/disco/store/fd_pending_slots.c [hunk 12]
```diff
@@ -73,127 +85,50 @@ fd_pending_slots_delete( void * pending_slots ) {
...
+        !fd_pending_slots_treap_fwd_iter_done( iter ); ) {
+    fd_pending_slots_treap_ele_t * ele = fd_pending_slots_treap_fwd_iter_ele( iter, pending_slots->pool );
```
Comment (lidatong) [inline src/disco/store/fd_pending_slots.c:127]: rename to prev

File: src/disco/store/fd_pending_slots.c [hunk 13]
```diff
@@ -73,127 +85,50 @@ fd_pending_slots_delete( void * pending_slots ) {
...
+    if ( ele->slot > lo_wmark ) break;
+
+    iter = fd_pending_slots_treap_fwd_iter_next( iter, pending_slots->pool );
```
Comment (lidatong) [inline src/disco/store/fd_pending_slots.c:130]: add a comment explaining what you're doing eg. "advance the iterator. it is safe to remove the previous element while iterating a treap <...> "

File: src/disco/store/fd_pending_slots.c (hunk 12 above)
Comment (lidatong) [reply src/disco/store/fd_pending_slots.c:127]: also no space after `for`, similar to `if`. see examples in kevin's code like here: https://github.com/firedancer-io/firedancer/blob/main/src/tango/mcache/fd_mcache.c#L65

File: src/disco/store/fd_pending_slots.c [hunk 14]
```diff
@@ -1,8 +1,7 @@
 #include "fd_pending_slots.h"
 
 void *
-fd_pending_slots_new( void * mem, ulong lo_wmark ) {
-
+fd_pending_slots_new( void * mem ) {
```
Comment (lidatong) [inline src/disco/store/fd_pending_slots.c:4]: `seed` should be passed in to initialize the rng

Comment (yhzhangjump) [reply src/disco/store/fd_pending_slots.c]: right, that's better, `seed` has been added

**Pull request closed without merging**